### DISCUSSION TITLE
2 traces generated instead of 1

### DISCUSSION THREAD

**Discussion by nithin1995** [langfuse/langfuse#2244] [Support] [answered]
I have a dataset and am running my solution against a dataset. For each item in the dataset, I do a get_langchain_handler(). This dataset is evaluated against a agent workflow, where several tools are used etc etc. 

I want to track both the function calls and the LLM calls. So I have added the required decorators and have langfuse handlers as callbacks in the LLM calls. The result of this is 2 traces, I would like to see everything captured by the decorator and the callback handler in a single trace, how do I do that?

**Comment by marcklingen** [comment]
@hassiebp is currently working on better support of decorated functions within a dataset run. The solution will be to have one parent functions that's decorated and associated with the dataset run. This decorated function then executes the langchain agent workflow and uses the [decorator/langchain interop](https://langfuse.com/docs/sdk/python/decorators#langchain).

**Reply by hassiebp** [reply]
Hi @nithin1995, we have just released v2.35.0 of the Python SDK that should help addressing your issue 🙂 Link to docs: https://langfuse.com/docs/datasets/overview#run-experiment-on-a-dataset

You can basically run dataset item executions with your main application function that does the regular tracing with the @observe decorator and the nested langfuse_context.get_current_langchain_handler() . The dataset item run will be automatically linked to that trace.

Let us know if this fixed it for you 🙏🏾

**Reply by nithin1995** [reply]
Thank you, I was able to test it with 2.35.0. I think I'm missing something, I still get 2 traces instead of 1. Let me give more details about my implementation, please let me know what needs to be changed to get a single trace instead of 2. 

```
def score_batch():
    for dataset_item in dataset:
        with dataset_item.observe(run_name) as trace_id:
            langfuse_handler = dataset_item.get_langchain_handler(run_name)
            result = run_llm_app(pipeline, question, batch, callback_handlers=[langfuse_hander])

@observe()
def run_llm_app():
```
    

Along with run_llm_app I have also decorated the tools it might use and have also decorated other functions that I want to trace which run_llm_app calls. This gets logged as 2 traces under the same dataset item - 
trace 1 - based on the observe decorator 
trace 2 - from the LLM calls where langfuse handler is passed

Is this expected? Can you help me?

**Reply by hassiebp** [reply, marked as answer]
Hi @nithin1995 - you can delete the line

``` python
            langfuse_handler = dataset_item.get_langchain_handler(run_name)
```

as I assume your `run_llm_app` function is already handling all langchain calls by getting a handler via `langfuse_context.get_current_langchain_handler` somewhere _inside_ the decorated function. Your implementation would become simply

```python
def score_batch():
    for dataset_item in dataset:
        with dataset_item.observe(run_name) as trace_id:
            result = run_llm_app(pipeline, question, batch)

@observe()
def run_llm_app(pipeline, question, batch):
    langfuse_handler = langfuse_context.get_current_langchain_handler()
    # This handler can be registered as a callback in all following Langchain executions (if not already done so)
```

**Reply by nithin1995** [reply]
It worked, thanks !
